### PR TITLE
controller: Fix adding default disk back after extra mount disk unmounted

### PR DESCRIPTION
**Pull request by boknowswiki**
This fix moves the multiple disks checking in the same filesystem ID after UUID checking.
If the UUID is not avaiable, we do not need to check the filesystem ID.

https://github.com/longhorn/longhorn/issues/1291

Signed-off-by: Bo Tao <bo.tao@rancher.com>

### CODE REVIEW

File: controller/node_controller.go [hunk 1]
```diff
@@ -637,14 +657,6 @@ func (nc *NodeController) syncDiskStatus(node *longhorn.Node) error {
 				diskUUID = diskConfig.DiskUUID
 			}
 			if diskStatusMap[id].DiskUUID == "" {
-				if len(disks) > 1 {
```
Comment (yasker) [inline controller/node_controller.go:640]: It seems this fix has an issue. On line 649, we will create the disk configuration which is a file on the disk. That should be after we've verified there is no filesystem conflict.

Now if we move the check to a later place, the disk config (you can find it on the disk as `longhorn-disk.cfg`) might be created by mistake on the filesystem the disk shouldn't be valid.

Comment (boknowswiki) [reply controller/node_controller.go:640]: @yasker ,
I have move this check before the UUID check. But one question about this `longhorn-disk.cfg`, there are some checking after this file created. And I also don't see any deletion if we error out. I assume, maybe we clean this directory then it got deleted?

Comment (yasker) [reply controller/node_controller.go:640]: 1. I think you're referring to schedulable storage calculation later? Those won't impact the disk itself.
1. For deletion, `generateDiskConfig` should be atomic. There are two functions executed after https://github.com/longhorn/longhorn-manager/blob/master/util/util.go#L936 . I think we should add cleanup for failure then.

File: util/util.go [hunk 2]
```diff
@@ -643,6 +643,33 @@ func CreateDiskPathReplicaSubdirectory(path string) error {
...
+	nsPath := iscsi_util.GetHostNamespacePath(HostProcPath)
+	nsExec, err := iscsi_util.NewNamespaceExecutor(nsPath)
+	if err != nil {
+		return err
+	}
+
+	dirPath := filepath.Join(path, ReplicaDirectory)
+	filePath := filepath.Join(path, DiskConfigFile)
+
+	// Check if the replica directory exist, delete it
```
Comment (yasker) [inline util/util.go:656]: Don't use `rm -rf`. It's not safe. Use `rmdir` instead. If it's created by us, it should be empty.

File: util/util.go [hunk 3]
```diff
@@ -933,6 +960,18 @@ func GenerateDiskConfig(path string) (*DiskConfig, error) {
 	if _, err := nsExec.Execute("ls", []string{filePath}); err == nil {
 		return nil, fmt.Errorf("disk cfg on %v exists, cannot override", filePath)
 	}
+
+	// Cleaning up disk cfg file and replica subdirectory if anything
+	// goes wrong
+	defer func() {
+		if err != nil {
+			if derr := DeleteDiskPathReplicaSubdirectoryAndDiskCfgFile(path); derr != nil {
```
Comment (yasker) [inline util/util.go:968]: nsExec can be passed in as a parameter for the function, so we don't need to create it again.

File: util/util.go [hunk 4]
```diff
@@ -643,6 +643,33 @@ func CreateDiskPathReplicaSubdirectory(path string) error {
 	return nil
 }
 
+func DeleteDiskPathReplicaSubdirectoryAndDiskCfgFile(path string) error {
```
Comment (yasker) [inline util/util.go:646]: This part of code change in util.go addresses a different issue, move it to a separate commit.

File: util/util.go [hunk 5]
```diff
@@ -933,6 +960,18 @@ func GenerateDiskConfig(path string) (*DiskConfig, error) {
 	if _, err := nsExec.Execute("ls", []string{filePath}); err == nil {
 		return nil, fmt.Errorf("disk cfg on %v exists, cannot override", filePath)
 	}
+
+	// Cleaning up disk cfg file and replica subdirectory if anything
+	// goes wrong
+	defer func() {
+		if err != nil {
+			if derr := DeleteDiskPathReplicaSubdirectoryAndDiskCfgFile(path); derr != nil {
+				err = errors.Wrapf(err, "Cleaning up disk config failed err %v with path %v", derr, path)
```
Comment (yasker) [inline util/util.go:969]: `err = errors.Wrapf(err, "cleaning up disk config path %v failed with error:  %v", path, derr)`

Always put related error at the end so it's easy to notice.

Also, don't capitalize the first character in error message, there can be other errors wrap on it too.

File: util/util.go [hunk 6]
```diff
@@ -933,6 +960,18 @@ func GenerateDiskConfig(path string) (*DiskConfig, error) {
 	if _, err := nsExec.Execute("ls", []string{filePath}); err == nil {
 		return nil, fmt.Errorf("disk cfg on %v exists, cannot override", filePath)
 	}
+
+	// Cleaning up disk cfg file and replica subdirectory if anything
```
Comment (yasker) [inline util/util.go:964]: Don't need to comment here. It's obvious in the code.

File: controller/node_controller.go [hunk 7]
```diff
@@ -571,6 +571,26 @@ func (nc *NodeController) getDiskInfoMap(node *longhorn.Node) map[string]*diskIn
 	return result
 }
 
+// Check multiple disks belongs to the same filesystem ID
+// Skip the ones are not in ready status
+func (nc *NodeController) checkSameFsidInDisks(id string, disks []string, diskStatusMap map[string]*types.DiskStatus) bool {
```
Comment (yasker) [inline controller/node_controller.go:576]: The function name is confusing. `check` doesn't say if we expect true or false if there is same FSID among the disks.

Use `isFSIDDuplicatedWithExistingReadyDisk(name string, ...)` instead. FSID is an abbreviation so it needs to be capitalized. Also, use `name` instead of `id`, since it can be mistaken for the `fsid` in this case. Same with the variable used in the loop.

Also, update the comment to indicate that all the disks passed in here have the same FSID.

File: controller/node_controller.go [hunk 8]
```diff
@@ -636,15 +656,24 @@ func (nc *NodeController) syncDiskStatus(node *longhorn.Node) error {
...
-						nc.eventRecorder, node, v1.EventTypeWarning)
-					continue
-				}
+						nc.eventRecorder, node,
+						v1.EventTypeWarning)
+				continue
+
+			}
+
+			if diskStatusMap[id].DiskUUID == "" {
```
Comment (yasker) [inline controller/node_controller.go:676]: Why move this line? If the disk already has UUID assigned, we can use logic on line 691 to validate the disk.

Comment (boknowswiki) [reply controller/node_controller.go:676]: @yasker sorry, I don't understand this comment, I moved the `if len(disks) > 1 {` out of `if diskStatusMap[id].DiskUUID == "" {`.

Comment (yasker) [reply controller/node_controller.go:676]: I meant, why you moved it? Previously we only check FSID if the UUID doesn't exist yet for the disk. Now we will apply FSID on every disk as the result of moving it, right?
